libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -644,25 +654,35 @@ string YulUtilFunctions::overflowCheckedUnsignedExpFunction()
 				if iszero(exponent) { power := 1 leave }
 				if iszero(base) { power := 0 leave }
 
-				power := 1
-
-				for { } gt(exponent, 1) {}
+				// Specializations for small bases
+				switch base
+				// 0 is handled above
+				case 1 { power := 1 leave }
+				case 2
 				{
-					// overflow check for base * base
-					if gt(base, div(max, base)) { revert(0, 0) }
-					if and(exponent, 1)
-					{
-						// no check needed here because base >= power
-						power := mul(power, base)
-					}
-					base := mul(base, base)
-					exponent := <shr_1>(exponent)
+					if gt(exponent, 255) { revert(0, 0) }
+					power := exp(2, exponent)
+					if gt(power, max) { revert(0, 0) }
+					leave
```

Comment:
We can compute an upper bound for the exponent here: it would be `min(255, boost::multiprecision::msb(max))`. So we can skip the last `gt(power, max) { revert(0, 0) }`

Reply:
Sounds good!

Reply:
Oh, but "max" is a runtime parameter, at least as it is written now....

Reply:
I think `max` is known at compile time. Maybe we should change the function signature and substitute for `max` in the template. Also the same applies for bases `10` and `306`. The `maxExponent` for those cases would be `std::min(77, msb(max))` and `std::min(31, msb(max))`.

Reply:
It is known at compile time, but making it a compile-time constant means that we have different functions for each value of max.

Reply:
But actually since the main loop is extracted, there is not much left anymore - will take a look!

Reply:
If we don't only want to have this for base two, we basically need https://github.com/ethereum/solidity/pull/9816/files#diff-743d8f2ce7665d50d4c151be906c2556R645 right?

Reply:
Ah yes, the `msb` formula only works for 2.

Reply:
I checked it again and think it's not worth it - it only saves a single comparison but adds lots of more or less duplicate code.

Reply:
What about implementing only for `base = 2`?

Reply:
The problem is that we still need a different function for each type.